compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.test.cpp

```diff
@@ -0,0 +1,401 @@
+    {
+      _back_perm->name("back_transpose/perm");
+      init_circle_const(_back_perm, back_perm);
+    }
+
+    _back_transpose = g()->nodes()->create<luci::CircleTranspose>();
+    {
+      _back_transpose->name("back_transpose");
+      _back_transpose->a(_mid_reshape);
+      _back_transpose->perm(_back_perm);
+    }
+
+    output()->from(_back_transpose);
+  }
+
+private:
+  luci::CircleTranspose *_front_transpose;
+  luci::CircleConst *_front_perm;
+
+  luci::CircleReshape *_mid_reshape;
+  luci::CircleConst *_mid_shape;
+
+  luci::CircleTranspose *_back_transpose;
+  luci::CircleConst *_back_perm;
+};
+
+} // namespace
+
+TEST(RemoveUnnecessaryTransposeNetPass, rank_reduction_pattern1)
+{
+  TransposeReshapeTransposeGraph g;
+  luci::RemoveUnnecessaryTransposeNetPass pass;
+
+  /**
+   * from efficient-former
+   *
+   * (1, 14, 14, 192)
+   *      A   B   C
+   *      |
+   * (1, 192, 14, 14)
+   *      C   A   B
+   *      |
+   * (1, 192, 196)
+   *      C   AB
+   *      |
+   * (1, 196, 192)
+   *      AB   C
+   */
+  g.init(/*in*/ {1, 14, 14, 192}, /*perm*/ {0, 3, 1, 2}, /*reshape*/ {1, 192, 196},
+         /*perm*/ {0, 2, 1}, /*out*/ {1, 196, 192});
+
+  EXPECT_TRUE(pass.run(g.g()));
+}
+
+TEST(RemoveUnnecessaryTransposeNetPass, rank_reduction_pattern2)
+{
+  TransposeReshapeTransposeGraph g;
+  luci::RemoveUnnecessaryTransposeNetPass pass;
+
+  /** from detr
+   *
+   * (1, 7, 7, 448)
+   *     A  B   C
+   *      |
+   * (1, 448, 7, 7)
+   *      C   A  B
+   *       |
+   * (1, 448, 49)
+   *      C   AB
+   *      |
+   * (1, 49, 448)
+   *     AB   C
+   */
+  g.init(/*in*/ {1, 7, 7, 448}, /*perm*/ {0, 3, 1, 2}, /*reshape*/ {1, 448, 49}, /*perm*/ {0, 2, 1},
+         /*out*/ {1, 49, 448});
+
+  EXPECT_TRUE(pass.run(g.g()));
+}
+
+TEST(RemoveUnnecessaryTransposeNetPass, rank_reduction_pattern3)
+{
+  TransposeReshapeTransposeGraph g;
+  luci::RemoveUnnecessaryTransposeNetPass pass;
+
+  /**
+   * (1, 100, 10, 12)
+   *      A   B   C
+   *        |
+   * (1, 10, 12, 100)
+   *     B   C   A
+   *        |
+   *    (120, 100)
+   *      BC   A
+   *        |
+   *    (100, 120)
+   *      A   BC
+   */
+  g.init(/*in*/ {1, 100, 10, 12}, /*perm*/ {0, 2, 3, 1}, /*reshape*/ {120, 100}, /*perm*/ {1, 0},
+         /*out*/ {100, 120});
+
+  EXPECT_TRUE(pass.run(g.g()));
+}
+
+// TODO Enable this pass after expand pass coverage
+TEST(RemoveUnnecessaryTransposeNetPass, DISABLED_rank_expand_pattern1)
+{
+  TransposeReshapeTransposeGraph g;
+  luci::RemoveUnnecessaryTransposeNetPass pass;
+
+  /**
+   * (10, 15, 20)
+   *  A   B   C
+   *      |           => transpose
+   * (10, 20, 15)
+   *  A   C   B
+   *      |           => reshape
+   * (10, 20, 3, 5)
+   *  A   C   Ba Bb
+   *      |           => transpose
+   * (10, 3, 5, 20)
+   *  A   Ba Bb C
+   */
+
+  g.init(/*in*/ {10, 15, 20}, /*perm*/ {0, 2, 1}, /*reshape*/ {10, 20, 3, 5}, /*perm*/ {0, 2, 3, 1},
+         /*out*/ {10, 3, 5, 20});
+
+  EXPECT_TRUE(pass.run(g.g()));
```

Comment:
As I understand `pass.run()` should return false as the graph doesn't meet the condition.
why is it true?

Reply:
`DISABLED` tests are written for the previous draft (which also cover rank-expand cases). :sweat_smile: 
That's why it expects `TRUE` and also `DISABLED` in this draft. 

I'd better remove(or revise) `DISABLED` tests for this draft. 
Thank you for pointing it out.

Reply:
ah... `DISABLED_` looks like gtest feature...
in (near) future, do you have plans to support expanding case?
if so, let's add them when the feature is supported.
optionally, as for now, we can add these cases to negative test cases with some TODO comments to support them in the future.

Reply:
> in (near) future, do you have plans to support expanding case?

I don't have a plan to expand the feature(coverage) in 'near'.

 > optionally, as for now, we can add these cases to negative test cases with some TODO comments to support them in the future. 

This looks good :+1:  Let me move `DISABLED` cases to negative test cases with some comments (e.g. rank-expansion pattern didn't meet the condition ... )

Reply:
On second thought... if we have to support rank-expansion pattern these NEG test cases should be removed then.
Which would be a opposite direction of our SE guide.